src/services/services.ts

```diff
@@ -4464,6 +4508,475 @@ namespace ts {
+                const result: VisibleModuleInfo[] = [];
+
+                if (host.readFile && host.fileExists) {
+                    for (const packageJson of findPackageJsons(scriptPath)) {
+                        const package = tryReadingPackageJson(packageJson);
+                        if (!package) {
+                            return;
+                        }
+
+                        const nodeModulesDir = combinePaths(getDirectoryPath(packageJson), "node_modules");
+                        const foundModuleNames: string[] = [];
+
+                        if (package.dependencies) {
+                            addPotentialPackageNames(package.dependencies, foundModuleNames);
+                        }
+                        if (package.devDependencies) {
+                            addPotentialPackageNames(package.devDependencies, foundModuleNames);
+                        }
+
+                        for (const moduleName of foundModuleNames) {
+                            const moduleDir = combinePaths(nodeModulesDir, moduleName);
+                            result.push({
+                                moduleName,
+                                moduleDir
+                            });
+                        }
+                    }
+                }
+
+                return result;
+
+                function tryReadingPackageJson(filePath: string) {
+                    try {
+                        const fileText = host.readFile(filePath);
+                        return JSON.parse(fileText);
+                    }
+                    catch (e) {
+                        return undefined;
+                    }
+                }
+
+                // Add all the package names that are not in the @types scope
+                function addPotentialPackageNames(dependencies: any, result: string[]) {
+                    for (const dep in dependencies) {
+                        if (dependencies.hasOwnProperty(dep) && !startsWith(dep, "@types/")) {
+                            result.push(dep);
+                        }
+                    }
+                }
+            }
+
+            function createCompletionEntryForModule(name: string, kind: string): ImportCompletionEntry {
+                return { name, kind, sortText: name };
+            }
+
+            // Replace everything after the last directory seperator that appears
+            // FIXME: do we care about the other seperator?
+            function getDirectoryFragmentTextSpan(text: string, textStart: number): TextSpan {
+                const index = text.lastIndexOf(directorySeparator);
+                const offset = index !== -1 ? index + 1 : 0;
+                return { start: textStart + offset, length: text.length - offset };
+            }
+
+            // Returns true if the path is explicitly relative to the script (i.e. relative to . or ..)
+            function isPathRelativeToScript(path: string) {

```

Comment:
Would `isExternalModuleNameRelative` (in utilities.ts) work in place of this function?